### PR TITLE
Pass static generator parameters to Prescient from bidder

### PR DESCRIPTION
**Pull request by xiangao1**
## Fixes
Pass static generator parameters to Prescient from bidder.

## Summary/Motivation:
This PR enables users to pass static generator parameters, e.g., Pmax, Pmin, to Prescient through double loop. 

First, I am proposing a class that holds the parameters. 

How the class will be used?
1. An instance of this class is required to create a `bidder` / `tracker`. (This can be part of the effort to address this: https://github.com/IDAES/idaes-pse/pull/787#discussion_r848854374)
2. When bidding into the market, these values will be passed to Prescient by `bidder`.
3. In this way, we don't have to maintain a lot of individual properties, e.g., pmin, pmax, in a `bidder` object.

Then I implemented functionalities to pass these static parameters using the model data object described above through the coordinator.

## Changes proposed in this PR:
- Implement the data structure and corresponding validators.
- Unit test for the data structure.
- Implement functionalities to pass static parameters into Prescient.
- Update tests and examples correspondingly.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
5. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by bknueven — COMMENTED** [review]
Given that this code is *doing* validating, I think it would be nice to have at least the `ValueError` exceptions tested.

This code also does floating point comparisons. I wonder if things like `value < self.min_value` need to be more like `value < self.min_value and not math.isclose(value, self.max_value` (e.g., "much less than" https://github.com/grid-parity-exchange/Egret/blob/828bd72236bced3a2c063bf63665fb569f4dfe3f/egret/model_library/unit_commitment/params.py#L1108)

If the validators need to be a bit more forgiving, e.g., because they're validating the output of an optimization model, it may be best for them to *return* the validated value so it can be adjusted, if needed, e.g., if `value < self.min_value` but `value` was close to `self.min_value` then `self.min_value` could be used instead.

**Comment by xiangao1** [comment]
> Given that this code is _doing_ validating, I think it would be nice to have at least the `ValueError` exceptions tested.
> 
> This code also does floating point comparisons. I wonder if things like `value < self.min_value` need to be more like `value < self.min_value and not math.isclose(value, self.max_value` (e.g., "much less than" https://github.com/grid-parity-exchange/Egret/blob/828bd72236bced3a2c063bf63665fb569f4dfe3f/egret/model_library/unit_commitment/params.py#L1108)
> 
> If the validators need to be a bit more forgiving, e.g., because they're validating the output of an optimization model, it may be best for them to _return_ the validated value so it can be adjusted, if needed, e.g., if `value < self.min_value` but `value` was close to `self.min_value` then `self.min_value` could be used instead.

@bknueven Thanks! I have added more tests and applied "much less than".

Right now I think this model data structure will only be used to store the static data. The user will "manually" create one instance of this, and then use it to create `bidder` and `tracker`. So I think it's ok that the validators are not forgiving. If the bid comes from optimization problem solutions, we have some code to adjust values, e.g., [here](https://github.com/IDAES/idaes-pse/blob/9abeb3bd34cc03c21120637cecb2d34ab1961bb0/idaes/apps/grid_integration/bidder.py#L359).

**Comment by xiangao1** [comment]
Hi @bknueven @adowling2, I have refactored this PR for passing static parameters from bidder to Prescient. Could you review the changes? Thanks!

File: idaes/apps/grid_integration/model_data.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+
+        if production_cost_bid_pairs[0][0] != self.p_min:
+            raise ValueError(
+                f"The first power output in the bid should be the Pmin {self.p_min}, but {production_cost_bid_pairs[0][0]} is provided."
+            )
+
+        if production_cost_bid_pairs[-1][0] != self.p_max:
+            raise ValueError(
+                f"The last power output in the bid should be the Pmax {self.p_max}, but {production_cost_bid_pairs[-1][0]} is provided."
+            )
```
Comment (bknueven) [inline idaes/apps/grid_integration/model_data.py:322]: I think its fine to do this validation, but its no longer required for Egret/Prescient.

File: idaes/apps/grid_integration/coordinator.py [hunk 2]
```diff
@@ -353,6 +356,64 @@ def _clone_tracking_model(self):
...
+        for param, value in self.bidder.bidding_model_object.model_data:
+            if param != "gen_name":
+                continue
+            elif param == "p_cost":
+                curve_value = convert_marginal_costs_to_actual_costs(value)
+                gen_dict[param] = {
+                    "data_type": "cost_curve",
+                    "cost_curve_type": "piecewise",
+                    "values": curve_value,
+                }
```
Comment (bknueven) [inline idaes/apps/grid_integration/coordinator.py:380]: Is this covered in an integration test? seems like important code to cover for Prescient interaction.

**Review by bknueven — COMMENTED** [review]
Looks pretty good -- I have a question about testing/coverage.

**Comment by ksbeattie** [comment]
@adowling2 is this needed for the DISPATCHES April/May (internal) or the June (milestone) release?

**Comment by adowling2** [comment]
This will be helpful for two of the DISPATCHES case studies. It is definitely nice to have for June; I am not sure if the case studies will need it.

File: idaes/apps/grid_integration/coordinator.py (hunk 2 above)
Comment (xiangao1) [reply idaes/apps/grid_integration/coordinator.py:380]: Thanks @bknueven ! I fixed it!

**Comment by ksbeattie** [comment]
If this is needed for the DISPATCHES June release, then it needs to get into the idaes-pse May release, so this month.

File: idaes/apps/grid_integration/coordinator.py [hunk 3]
```diff
@@ -353,6 +356,64 @@ def _clone_tracking_model(self):
...
+                if "p_fuel" in gen_dict:
+                    gen_dict.pop("p_fuel")
+            else:
+                gen_dict[param] = value
```
Comment (adowling2) [inline idaes/apps/grid_integration/coordinator.py:384]: Add a comment here explaining what this is doing?

File: idaes/apps/grid_integration/coordinator.py [hunk 4]
```diff
@@ -353,6 +356,64 @@ def _clone_tracking_model(self):
...
+
+            if param == "startup_cost" and "startup_fuel" in gen_dict:
```
Comment (adowling2) [inline idaes/apps/grid_integration/coordinator.py:386]: Is there a reason this is outside the if, elif, else statement above? If so, just add a one-line comment explaining

**Review by adowling2 — CHANGES_REQUESTED** [review]
Small request for small explanation/extra comments.

**Comment by xiangao1** [comment]
Thanks, @adowling2 ! I have addressed your comments.

**Review by adowling2 — APPROVED** [review]
Looks good; my earlier comments have been addressed.

**Comment by xiangao1** [comment]
@ksbeattie @lbianchi-lbl This is ready to merge.

**Pull request merged**